### PR TITLE
Added VerifyClusterOperatorsVersion func

### PR DESCRIPTION
**Pull request by siserafin**
Added function to verify all the clusteroperators version after the ocp upgrade

### CODE REVIEW

File: pkg/clusteroperator/list.go [hunk 1]
```diff
@@ -131,3 +131,50 @@ func WaitForAllClusteroperatorsStopProgressing(
 
 	return false, err
 }
+
+// VerifyClusterOperatorsVersion checks if all the clusterOperators have version desiredVersion.
```
Comment (kononovn) [inline pkg/clusteroperator/list.go:135]: Can you re-factor and simplify this function?

**Review by kononovn — CHANGES_REQUESTED** [review]
Please check comment above.

File: pkg/clusteroperator/list.go [hunk 2]
```diff
@@ -131,3 +131,50 @@ func WaitForAllClusteroperatorsStopProgressing(
 
 	return false, err
 }
```
Comment (kononovn) [inline pkg/clusteroperator/list.go:133]: Please change commit message to:
clusteroperator: Added VerifyClusterOperatorsVersion func

Comment (siserafin) [reply pkg/clusteroperator/list.go:133]: done, thanks

File: pkg/clusteroperator/list.go (hunk 1 above)
Comment (siserafin) [reply pkg/clusteroperator/list.go:135]: done, thanks

File: pkg/clusteroperator/clusteroperator.go [hunk 3]
```diff
@@ -223,6 +223,41 @@ func (builder *Builder) WaitUntilConditionTrue(
 		})
 }
 
+// HasDesiredVersion checks if an operator has a desiredVersion.
+func (builder *Builder) HasDesiredVersion(desiredVersion string) (bool, error) {
```
Comment (kononovn) [inline pkg/clusteroperator/clusteroperator.go:227]: Let's simplify it a bit more. Please review code example below:


```
// HasDesiredVersion checks if an operator has a desiredVersion.
func (builder *Builder) HasDesiredVersion(desiredVersion string) (bool, error) {
	if valid, err := builder.validate(); !valid {
		return false, err
	}

	if desiredVersion == "" {
		return false, fmt.Errorf("desiredVersion can't be empty")
	}
	if builder.Object.Status.Versions == nil {
		return false, fmt.Errorf("undefined cluster operator status versions")
	}
	
	for _, operatorVersion := range builder.Object.Status.Versions {
		glog.V(100).Infof("Testing %s and %s", operatorVersion.Version, desiredVersion)

		if operatorVersion.Version == desiredVersion {
			glog.V(100).Infof("The clusterOperator %s version matches the desired version %s",
				builder.Definition.Name, desiredVersion)

			return true, nil
		}

		glog.V(100).Infof(
			"The clusterOperator %s doesn't have the desired version %s, but has version %s. Continue lookup",
			builder.Definition.Name, desiredVersion, operatorVersion.Version)
	}

	glog.V(100).Infof("lookup failed.The clusterOperator %s doesn't have the desired version %s",
		builder.Definition.Name, desiredVersion)

	return false, nil
}
```

File: pkg/clusteroperator/clusteroperator_test.go [hunk 4]
```diff
@@ -148,12 +149,123 @@ func TestClusterOperatorGet(t *testing.T) {
 	}
 }
 
+func TestHasDesiredVersion(t *testing.T) {
```
Comment (kononovn) [inline pkg/clusteroperator/clusteroperator_test.go:152]: Please re-name to:
**TestClusterOperatorHasDesiredVersion**

File: pkg/clusteroperator/clusteroperator_test.go [hunk 5]
```diff
@@ -148,12 +149,123 @@ func TestClusterOperatorGet(t *testing.T) {
...
+		},
+	}
+	for _, testCase := range testCases {
+		result, err := testCase.testClusterOperator.HasDesiredVersion(testCase.desiredVersion)
+		assert.Equal(t, testCase.expectedOutput, result)
+		assert.Equal(t, testCase.expectedError, err)
+	}
+}
+
+func TestVerifyClusterOperatorsVersion(t *testing.T) {
```
Comment (kononovn) [inline pkg/clusteroperator/clusteroperator_test.go:212]: Please add new file **list_test.go** and move this function it the new file.

File: pkg/clusteroperator/clusteroperator_test.go [hunk 6]
```diff
@@ -196,3 +374,15 @@ func newBuilder(apiClient *clients.Settings, name string) *Builder {
 
 	return builder
 }
+
+// createFakeObject method creates a new fake instance of builder
+// and initializes the builder Object with its definition.
+func createFakeObject(apiClient *clients.Settings, name string, status configV1.ClusterOperatorStatus) *Builder {
```
Comment (kononovn) [inline pkg/clusteroperator/clusteroperator_test.go:380]: Please remove this fucntion.

File: pkg/clusteroperator/clusteroperator_test.go [hunk 7]
```diff
@@ -172,17 +284,83 @@ func buildDummyClusterOperatorConfig() []runtime.Object {
...
+}
+
+// buildFakeClusterOperatorWithVersions creates a fake clusterOperator with
+// a specific list of ClusterOperatorStatusVersions for testing purpose.
+func buildFakeClusterOperatorWithVersions(apiClient *clients.Settings, versions []string) *Builder {
+	clusterOperatorStatusVersions := createFakeClusterOperatorStatusVersions(versions)
+
+	clusterOperatorStatus := createFakeClusterOperatorStatus(clusterOperatorStatusVersions)
+
+	clusterOperator := createFakeObject(apiClient, defaultClusterOperatorName, clusterOperatorStatus)
```
Comment (kononovn) [inline pkg/clusteroperator/clusteroperator_test.go:314]: Just refactor to:

```
	clusterOperator := newBuilder(apiClient, defaultClusterOperatorName, clusterOperatorStatus)
	clusterOperator.Object = clusterOperator.Definition
```

File: pkg/clusteroperator/clusteroperator_test.go [hunk 8]
```diff
@@ -172,17 +284,83 @@ func buildDummyClusterOperatorConfig() []runtime.Object {
...
+
+	return clusterOperator
+}
+
+// createFakeOperandVersion creates a fake OperandVersion.
+func createFakeOperandVersion(version string, name string) configV1.OperandVersion {
```
Comment (kononovn) [inline pkg/clusteroperator/clusteroperator_test.go:320]: Please delete this function

File: pkg/clusteroperator/clusteroperator_test.go [hunk 9]
```diff
@@ -172,17 +284,83 @@ func buildDummyClusterOperatorConfig() []runtime.Object {
...
+	return configV1.OperandVersion{Name: name, Version: version}
+}
+
+// createFakeClusterOperatorStatusVersions creates a fake list of ClusterOperatorStatusVersions.
+func createFakeClusterOperatorStatusVersions(versions []string) []configV1.OperandVersion {
+	var clusterOperatorStatusVersions []configV1.OperandVersion
+
+	for _, version := range versions {
+		operandVersion := createFakeOperandVersion(version, defaultClusterOperatorName)
+		clusterOperatorStatusVersions = append(clusterOperatorStatusVersions, operandVersion)
```
Comment (kononovn) [inline pkg/clusteroperator/clusteroperator_test.go:330]: use 
configV1.OperandVersion{Name: defaultClusterOperatorName, Version: version} 
instead operandVersion

**Review by kononovn — CHANGES_REQUESTED** [review]
Please fix few comments.

File: pkg/clusteroperator/clusteroperator_test.go (hunk 7 above)
Comment (siserafin) [reply pkg/clusteroperator/clusteroperator_test.go]: Done, thanks

File: pkg/clusteroperator/clusteroperator_test.go (hunk 6 above)
Comment (siserafin) [reply pkg/clusteroperator/clusteroperator_test.go]: Removed, thanks

File: pkg/clusteroperator/clusteroperator_test.go (hunk 5 above)
Comment (siserafin) [reply pkg/clusteroperator/clusteroperator_test.go]: Added, thanks

File: pkg/clusteroperator/clusteroperator_test.go (hunk 4 above)
Comment (siserafin) [reply pkg/clusteroperator/clusteroperator_test.go]: Re-named, thanks

File: pkg/clusteroperator/clusteroperator.go (hunk 3 above)
Comment (siserafin) [reply pkg/clusteroperator/clusteroperator.go:227]: Updated, thanks

**Review by kononovn — APPROVED** [review]
/lgtm

**Review by trewest — APPROVED** [review]
/lgtm

**Pull request merged**